ryhope/src/storage/memory.rs

```diff
@@ -371,35 +371,186 @@ where
+            (
+                EpochMapItem::Complete(first_user_epoch, _),
+                EpochMapItem::PartialUser(second_user_epoch),
+            ) => Some(first_user_epoch.cmp(&second_user_epoch)),
+            (
+                EpochMapItem::Complete(_, first_incremental_epoch),
+                EpochMapItem::PartialIncremental(second_incremental_epoch),
+            ) => Some(first_incremental_epoch.cmp(&second_incremental_epoch)),
+            (
+                EpochMapItem::Complete(first_user_epoch, first_incremental_epoch),
+                EpochMapItem::Complete(second_user_epoch, second_incremental_epoch),
+            ) => {
+                let user_epoch_cmp = first_user_epoch.cmp(&second_user_epoch);
+                let incremental_epoch_cmp = first_incremental_epoch.cmp(&second_incremental_epoch);
+                assert_eq!(
+                    user_epoch_cmp, incremental_epoch_cmp,
+                    "Breaking invariant of `EpochMapper`: both `UserEpoch` and `IncrementalEpoch`
+                    must be monotonically increasing"
+                );
+                Some(user_epoch_cmp)
+            }
+            _ =>
+            // all other cases are partial `EpochMapItem`s, which are never compared
+            {
+                None
+            }
+        }
+    }
+}
+
+impl Ord for EpochMapItem {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        // Safe since by construction we will never compare 2 Partial
+        // `EpochMapItem`s, which is the only case when `PartialOrd` is
+        // undefined
+        self.partial_cmp(other).unwrap()
+    }
+}
+
 #[derive(Clone, Debug)]
-pub struct InMemoryEpochMapper(BTreeMap<UserEpoch, IncrementalEpoch>);
+/// In-memory implementation of an `EpochMapper`, which allows to map a `UserEpoch`
+/// to an `IncrementalEpoch` used by storages. It basically handles two types of
+/// mappings, depending on how the epoch maps are inserted:
+/// - If the `UserEpoch`s being inserted are all incrementals, starting from an
+///   initial offset, then an optimized implementation is employed for this conversion
+/// - Otherwise, there is a more generic implementation that can handle any monotonically
+///   increasing sequence of `UserEpoch`s
+/// The first implementation is used until the `UserEpoch`s being inserted followed the
+/// incremental pattern; as soon as a non-incremental `UserEpoch` is inserted, then the
+/// implementation falls back to the more generic generic implementation
+pub struct InMemoryEpochMapper {
+    // Generic implementation to map monotonically increasing `UserEpoch`s to `IncrementalEpoch`s
+    generic_map: BTreeSet<EpochMapItem>,
+    // Optimized implemenetation for incremental `UserEpoch`s: it is sufficient to
+    // simply store:
+    // - The initial offset to convert between `UserEpoch`s and `IncrementalEpoch`s
+    // - The last inserted `UserEpoch`
+    incremental_epochs_map: Option<(UserEpoch, UserEpoch)>,
```

Comment:
Could you please convert the `(UserEpoch, UserEpoch)` to `struct Anchor (or whatever) { offset: UserEpoch, last_inserted: UserEpoch }`?

Reply:
Done in commit https://github.com/Lagrange-Labs/mapreduce-plonky2/pull/425/commits/9a606e9c9e8b321fc79145d50371e345d336870a